File: LC-1315/LC-1315.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    /**
     * Concepts: Depth-First Search (DFS)
     */
    int sumEvenGrandparent(TreeNode* root) {
        if (!root) return 0;
        
        int sums = 0;
        stack<TreeNode*> stk;
        stk.push(root);
        while (!stk.empty()) {
            TreeNode* curr = stk.top();
            stk.pop();
            
            if (curr->val % 2 == 0) {
                if (curr->left && curr->left->left)
                    sums += curr->left->left->val;
                if (curr->left && curr->left->right)
                    sums += curr->left->right->val;
                if (curr->right && curr->right->left)
                    sums += curr->right->left->val;
                if (curr->right && curr->right->right)
                    sums += curr->right->right->val;
            }
            
            if (curr->left)
                stk.push(curr->left);
            if (curr->right)
                stk.push(curr->right);
        }
        
        return sums;
    }
};

File: LC-1315/LC-1315.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumEvenGrandparent = function(root) {
    /**
     * Concepts: Depth-First Search (DFS)
     */

    if (!root) return 0;
    
    let sums = 0, stack = [];
    stack.push(root);
    while (stack.length) {
        let curr = stack[stack.length - 1];
        stack.pop();
        
        if (curr.val % 2 === 0) {
            if (curr.left && curr.left.left)
                sums += curr.left.left.val;
            if (curr.left && curr.left.right)
                sums += curr.left.right.val;
            if (curr.right && curr.right.left)
                sums += curr.right.left.val;
            if (curr.right && curr.right.right)
                sums += curr.right.right.val;
        }
        
        if (curr.left)
            stack.push(curr.left)
        if (curr.right)
            stack.push(curr.right)
    }
    
    return sums;
};